src/webview/columns/data-column.tsx

```diff
@@ -31,67 +55,240 @@ export class DataColumn implements ColumnContribution {
-        return groups;
-    }
 
-    protected renderWord(memory: Memory, options: TableRenderOptions, currentAddress: bigint): React.ReactNode {
-        const initialOffset = toOffset(memory.address, currentAddress, options.bytesPerWord * 8);
-        const finalOffset = initialOffset + options.bytesPerWord;
-        const bytes: React.ReactNode[] = [];
-        for (let i = initialOffset; i < finalOffset; i++) {
-            bytes.push(this.renderEightBits(memory, currentAddress, i));
-        }
-        this.applyEndianness(bytes, options);
-        return <span className='single-word' key={currentAddress.toString(16)}>{bytes}</span>;
+        if (words.length) { byteGroups.push({ startAddress, endAddress: range.endAddress, words }); }
+
+        return byteGroups;
     }
 
     protected applyEndianness<T>(group: T[], options: TableRenderOptions): T[] {
         // Assume data from the DAP comes in Big Endian so we need to revert the order if we use Little Endian
         return options.endianness === Endianness.Big ? group : group.reverse();
     }
+}
+
+export interface EditableDataColumnGroupProps {
+    range: BigIntMemoryRange;
+    group: DataColumnByteGroup;
+    options: TableRenderOptions;
+    writeMemory?(writeArguments: DebugProtocol.WriteMemoryArguments): Promise<void>;
+}
+
+export interface EditableDataColumnGroupState {
+    value: string;
+    isEdit: boolean;
+}
+
+export class EditableDataColumnGroup extends React.Component<EditableDataColumnGroupProps, EditableDataColumnGroupState> {
+
+    constructor(props: EditableDataColumnGroupProps) {
+        super(props);
+
+        this.state = {
+            isEdit: false,
+            value: this.asLine(this.props.group)
+        };
+    }
+
+    protected byteGroupStyle: React.CSSProperties = {
+        marginRight: `${DataColumn.Styles.MARGIN_RIGHT_PX}px`
+    };
+
+    protected get renderableCharacters(): number {
+        return this.props.options.bytesPerWord * this.props.options.wordsPerGroup * 2;
+    }
 
-    protected renderEightBits(memory: Memory, currentAddress: bigint, offset: number): React.ReactNode {
-        const { content, className, style, title } = this.getBitAttributes(memory, currentAddress, offset);
+    render(): React.ReactNode {
+        if (this.state.isEdit) {
+            return this.renderEditingGroup();
+        }
+
+        return this.renderReadonlyGroup();
+    }
+
+    componentDidUpdate(_prevProps: Readonly<EditableDataColumnGroupProps>, prevState: Readonly<EditableDataColumnGroupState>): void {
+        if (prevState.isEdit === false && this.state.isEdit) {
+            this.setState(prev => ({ ...prev, value: this.asLine(this.props.group) }));
+        }
+    }
+
+    protected asLine(group: DataColumnByteGroup): string {
+        return group.words.flatMap(w => w.bits.flatMap(b => b.value).join('')).join('');
+    }
+
+    protected renderEditingGroup(): React.ReactNode {
```

Comment:
In the Theia version of this functionality, the editable unit was the byte. I think in the current version, the equivalent would be the word. The group is mostly a unit of visual organization without a connection to the underlying memory units, so I'm not sure it makes sense as a unit of editing.

Reply:
`word` (we are discussing a new name for this in the scope of #100 , expect a proposal hopefully today) is for byte-addressable architectures always 1 Byte.

However, the Arm architecture works with different representations for different instructions. Hence, for such use cases, `group` would be the means to allow representations of larger data objects. For example BYTE (1 byte), HalfWORD (2 bytes), WORD (4 bytes) for an Arm M-profile architecture. As a SW developer I may want to see larger scalar representation, for example 64-bit when looking at a C `long long` on 32-bit machine. Especially in combination with the variables decorator column.

Also, if you group bytes/words to be displayed as one and separated by spaces, the IMO natural way of editing is to allow this on such a visual unit. That's at least how debuggers do it that I've worked with so far.

@colin-grant-work , how do you use `groups` in your use cases?

Reply:
I suppose that in general, I prefer operating at a level that is designed to represent _something_ in the underlying memory architecture, and I'm nervous about trying to make higher-level units meaningful in a general way when that meaning is very variable: groups can be almost arbitrarily large, and the 'meaningful' data unit depends on exactly what kind of data type people are dealing with. I don't think I'd want to encourage people to change their group size according to exactly the data type they want to edit (this struct is 10 bytes (which we can't have as a group size), but this one is 64 bytes (quite the chore and feat of memory to enter all in one go), etc. etc. - but how much is just padding, anyway?). The Theia version's solution for this was to allow keyboard navigation in the table and batching of multiple byte-level edits, which in my view would be the more natural way to achieve higher-level editability.

Reply:
> I suppose that in general, I prefer operating at a level that is designed to represent something in the underlying memory architecture, and I'm nervous about trying to make higher-level units meaningful in a general way when that meaning is very variable: groups can be almost arbitrarily large, and the 'meaningful' data unit depends on exactly what kind of data type people are dealing with. I don't think I'd want to encourage people to change their group size according to exactly the data type they want to edit (this struct is 10 bytes (which we can't have as a group size), but this one is 64 bytes (quite the chore and feat of memory to enter all in one go), etc. etc. - but how much is just padding, anyway?).

Out intention is to support SW developers in inspecting memory while debugging an application. I appreciate that we can't and shouldn't introduce grouping by more complex data types like structs. Apologies if above comment was misleading in that sense. We are rather thinking here of primitive data types resolving to a power-of-two bytes (char, short, int, long, long long, etc). Which helps already with a lot of use cases.
Such use cases could be that you'd like to inspect a section of stack the CPU has pushed addresses and values to. Or an array consisting of primitive data values, for example a ring buffer used by a peripheral or by one of your algorithms. In that case, the approach of bringing "meaning" into a group helps you to spot values which look odd. And you can quickly experiment with a change of such a meaningful value.

> The Theia version's solution for this was to allow keyboard navigation in the table and batching of multiple byte-level edits, which in my view would be the more natural way to achieve higher-level editability.

I have to admit that the approach of editing byte by byte and clicking an apply button as the Theia version requires was uncommon to me at the time I first used it. And it felt less efficient when being used to what other embedded debuggers do. Which is the approach we try to model with this PR: edit a value possibly larger than a byte and directly write to memory when you confirm by hitting enter. You still can navigate through the selected group with the arrow keys and modify byte by byte.

I've got a feeling I still don't fully understand how grouping is used by your users. But I'd sincerely love to. Can you give some examples of its usage?

Reply:
I think the difference is likely to lie more in philosophy or temperament than use-case, and I'm a bit conservative in this area: I _know_ that byte/word (depending on our resolution of the terminology confusion around those terms :-)) is a meaningful unit, at least as far as the chip running the code is concerned, but I _don't_ know that `group` is a meaningful unit, though it very well may be in a given context to a given user. Perhaps I can imagine a user complaining in our support channel "I use `group` to identify a struct of X type, and I only want to edit one field of that struct, but when I start editing, it turns the whole struct into an input." That is, I can imagine an incongruity between a particular user's use of `group` - there's no prescription about what `group` should mean, after all, and it doesn't reflect a base reality that everyone is subject to - and I'm averse to design decisions that might introduce such an incongruity. On the other hand, every change above the byte/word level consists of some changes at the byte/word level, and everyone is aware of the existence of a minimum addressable unit, so that the choice to focus on that level, while it may introduce slight inconvenience (or greater convenience, if the user's editing target is smaller than their `group` unit), it can't run against anyone's understanding of how memory works.

But that really is a philosophical stance and/or a defensive posture against possible objections, and it certainly isn't the only stance one can take, so let's talk about how things should work, especially in the context of:
 - The 'set value' option for variables.
 - The 'apply memory' functionality that @martin-fleck-at has introduced
 - Other debuggers / memory editors that we're familiar with

Reply:
A compromise position might be to allow editing by selection: allow the user to select what they want to edit, then enter editing mode, perhaps with a default to allow a user to click and type without selection. That largely removes our design decisions as an obstacle to users getting whatever they want.

Reply:
Another alternative: retain byte orientation, but automatically advance if the user continues entering values, like some two-factor authentication inputs, where each 'box' accepts a limited number of characters, but the cursor proceeds to the next box when appropriate.

Reply:
I get your point that users may want to edit with finer granularity. Yet I still think that, leaving any kind of meaning aside for a minute, users of our existing tools may be more used to selecting something that is displayed "as one". Which is the space-separated group.

Advancing from edit selection to edit selection has it's charm. But I see this as a future enhancement outside this PR. It could become part of a well-thought-through keyboard navigation (#106). 

Also, the "edit by selection" could be an interesting enhancement for the future. Regardless of the outcome of this discussion.

While I am starting to get worried about the complexity of our settings, I think this is a case where we need to let the user decide. Could be an option to specify the edit mode by `word` or `group` (or what we come up with in the separate terminology discussion). I would however keep this setting global. As you say, this may be a matter of personal taste, philosophy, or temperament. And I wouldn't expect that this needs to be configured per window instance.

TBF: I am keen to finalize this feature as is and work towards a release so we can gather real user feedback. Maybe we then get told right-away that the edit by `group` isn't very useful and can go back to `word` based editing without the need to add an setting. :-)

Reply:
Happy to go with editing at group level for now. I'm experimenting with edit by selection - I'll post the branch when I get it working.